feature-cancel-insurance/src/main/kotlin/com/hedvig/android/cancelinsurance/CancelInsuranceViewModel.kt

```diff
@@ -0,0 +1,124 @@
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.remember
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import arrow.core.Either
+import com.hedvig.android.cancelinsurance.data.CancelInsuranceUseCase
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.WhileSubscribed
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import kotlinx.datetime.Clock
+import kotlinx.datetime.LocalDateTime
+import kotlinx.datetime.TimeZone
+import kotlinx.datetime.atStartOfDayIn
+import kotlinx.datetime.toInstant
+import kotlinx.datetime.toLocalDateTime
+import kotlin.time.Duration.Companion.seconds
+
+@OptIn(ExperimentalMaterial3Api::class)
+internal class CancelInsuranceViewModel(
+  private val insuranceId: InsuranceId,
+  private val cancelInsuranceUseCase: CancelInsuranceUseCase,
+  clock: Clock = Clock.System,
+) : ViewModel() {
+  private val dateSubmissionError = MutableStateFlow(false)
+  private val dateSubmissionSuccess = MutableStateFlow(false)
+  private val isPerformingCancelInsuranceNetworkRequest = MutableStateFlow(false)
+
+  private val datePickerConfiguration = DatePickerConfiguration(clock)
+  val dateValidator = datePickerConfiguration.dateValidator
+
+  val uiState: StateFlow<CancelInsuranceUiState> = combine(
+    dateSubmissionError,
+    dateSubmissionSuccess,
+    isPerformingCancelInsuranceNetworkRequest,
+  ) { dateSubmissionError, dateSubmissionSuccess, isPerformingCancelInsuranceNetworkRequest ->
+    CancelInsuranceUiState(
+      datePickerState = datePickerConfiguration.datePickerState,
+      dateSubmissionError = dateSubmissionError,
+      dateSubmissionSuccess = dateSubmissionSuccess,
+      isLoading = isPerformingCancelInsuranceNetworkRequest,
+    )
+  }.stateIn(
+    viewModelScope,
+    SharingStarted.WhileSubscribed(5.seconds),
+    CancelInsuranceUiState(
+      datePickerConfiguration.datePickerState,
+      dateSubmissionError.value,
+      dateSubmissionSuccess.value,
+      isPerformingCancelInsuranceNetworkRequest.value,
+    ),
+  )
+
+  fun showedError() {
+    dateSubmissionError.update { false }
+  }
+
+  fun submitSelectedDate() {
+    val viewState = uiState.value
+    if (!viewState.canSubmitSelectedDate()) return
+    val selectedDateMillis = viewState.datePickerState.selectedDateMillis ?: return
+    isPerformingCancelInsuranceNetworkRequest.update { true }
+    viewModelScope.launch {
+      val result = cancelInsuranceUseCase.invoke(
+        insuranceId,
+        selectedDateMillis,
+      )
+      when (result) {
+        is Either.Left -> dateSubmissionError.update { true }
+        is Either.Right -> dateSubmissionSuccess.update { true }
+      }
+      isPerformingCancelInsuranceNetworkRequest.update { false }
```

Comment:
In cases like this, wont you get two state emissions - first with the error/success update, and then with the network request update? Would that be a problem?

Reply:
Yes this would trigger twice, and there would be an in-between where you'd get a uiState with (error = true or success = true) while loading is still true.

In case of the error, this'd mean that nothing changes in this UI, as it has no loading animation. If there was one, you'd get 1 frame where you'd get the snackbar while the loading is still showing.
In case of success it doesn't matter since success would mean you're navigating away directly, so you won't see the difference.

All that said, I did make this back into a single StateFlow in [this commit](https://github.com/HedvigInsurance/android/commit/e38151f6e0cb8d12b93a4faf49ee44e3a1790e93) 😊

The difference now is that we're not using `.stateIn()` which means that when we exit the app and go to the background, this StateFlow will keep alive forever until the app is killed vs how before with `.stateIn()` it was stopping this flow after 5 seconds of inactivity. Not a big deal in our case and in this screen, so let's go with it. If this turns out to be important in some other screen, where we have some hot flow ongoing then we'd need to figure out how to model this with `.stateIn()` while also not getting this double emission.